test/mocking_utils.hpp

```diff
@@ -0,0 +1,415 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef MOCKING_UTILS_HPP_
+#define MOCKING_UTILS_HPP_
+
+#include <stdio.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <time.h>
+
+#if !defined(_WIN32)
+#if defined(__MACH__)
+#include <mach/clock.h>
+#include <mach/mach.h>
+#endif
+#include <sys/types.h>
+#include <dirent.h>
+#endif
+
+#include <map>
+#include <string>
+
+#include "mimick/mimick.h"
+
+#include "rcutils/macros.h"
+
+template<typename T>
+class unique_in_scope
+{
+protected:
+  static T * instance;
+
+  explicit unique_in_scope(T * self)
+  {
+    prev_instance_ = instance;
+    instance = self;
+  }
+
+  ~unique_in_scope()
+  {
+    instance = prev_instance_;
+  }
+
+private:
+  T * prev_instance_ {nullptr};
+};
+
+template<typename T>
+T * unique_in_scope<T>::instance {nullptr};
+
+#if !defined(_WIN32)
+#if defined(__MACH__)
+
+class fake_clock : unique_in_scope<fake_clock>
+{
+public:
+  fake_clock()
+  : unique_in_scope(this)
+  {
+    mock_ = mmk_mock(RCUTILS_STRINGIFY(clock_get_time) "@lib:rcutils", clock_get_time_mock_t);
```

Comment:
What's the purpose of RCUTILS_STRINGIFY here instead of just writing out the full string?

Reply:
This is a fun detail. System libraries have a good deal of versioning mechanisms, and sometimes the binary API standards (symbol names) do not match the source API standards (identifier names). I've seen two techniques so far: that of using macros for the forward facing API, and that of using function in-lining. This is a counter-measure for the first, I have yet to find one for the later (reason why `glibc`'s `__xstat` is mentioned explicitly below).